perf: Faster decimal precision overflow checks

andygrove opened the pull request:
# Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

N/A

# Rationale for this change
 
<!--
Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.
-->

Small performance optimization.


```
null_if_overflow_precision_128
                        time:   [14.371 µs 14.403 µs 14.436 µs]
                        change: [-80.547% -80.475% -80.399%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild

null_if_overflow_precision_256
                        time:   [26.421 µs 26.531 µs 26.642 µs]
                        change: [-87.653% -87.587% -87.522%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

validate_decimal_precision_128
                        time:   [74.056 ns 74.401 ns 74.752 ns]
                        change: [-11.615% -10.563% -9.4441%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) low mild

validate_decimal_precision_256
                        time:   [215.61 ns 216.62 ns 217.68 ns]
                        change: [-0.8629% -0.3349% +0.1701%] (p = 0.22 > 0.05)
                        No change in performance detected.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high severe
```

# What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->


I noticed two areas of overhead in the current approach to verifying decimal precision.

1. In the event of an overflow, we are building a formatted string for an error message, which is then discarded in some cases, so we should avoid that cost
2. The range check code introduces a memcpy which can be avoided

I tested the following variations of the decimal precision check in Rust playground.

```rust
fn validate_decimal_precision1(value: i128, precision: u8) -> bool {
    if precision > DECIMAL128_MAX_PRECISION {
        return false;
    }
    let idx = usize::from(precision) - 1;
    value >= MIN_DECIMAL_FOR_EACH_PRECISION[idx] && value <= MAX_DECIMAL_FOR_EACH_PRECISION[idx]
}

// based on arrow-rs version
fn validate_decimal_precision2(value: i128, precision: u8) -> bool {
    if precision > DECIMAL128_MAX_PRECISION {
        return false;
    }

    let max = MAX_DECIMAL_FOR_EACH_PRECISION[usize::from(precision) - 1];
    let min = MIN_DECIMAL_FOR_EACH_PRECISION[usize::from(precision) - 1];

    if value > max {
        false
    } else if value < min {
        false
    } else {
        true
    }
}
```

`validate_decimal_precision1` avoids a `memcpy` that appears in `validate_decimal_precision2`:

```
playground::validate_decimal_precision1:
	subq	$1304, %rsp
	movb	%dl, %al
	movb	%al, 23(%rsp)
	movq	%rsi, 24(%rsp)
	movq	%rdi, 32(%rsp)
	movq	%rdi, 1264(%rsp)
	movq	%rsi, 1272(%rsp)
	movb	%al, 1287(%rsp)
	cmpb	$38, %al
	ja	.LBB9_2
	movb	23(%rsp), %al
	movb	%al, 1303(%rsp)
	movzbl	%al, %eax
	movq	%rax, %rcx
	subq	$1, %rcx
	movq	%rcx, 8(%rsp)
	cmpq	$1, %rax
	jb	.LBB9_4
	jmp	.LBB9_3

playground::validate_decimal_precision2:
	subq	$1368, %rsp
	movb	%dl, %al
	movb	%al, 55(%rsp)
	movq	%rsi, 56(%rsp)
	movq	%rdi, 64(%rsp)
	movq	%rdi, 1296(%rsp)
	movq	%rsi, 1304(%rsp)
	movb	%al, 1327(%rsp)
	cmpb	$38, %al
	ja	.LBB10_2
	leaq	80(%rsp), %rdi
	leaq	.L__unnamed_5(%rip), %rsi
	movl	$608, %edx
	callq	memcpy@PLT   <----- MEMCPY HERE
	movb	55(%rsp), %al
	movb	%al, 1367(%rsp)
	movzbl	%al, %eax
	movq	%rax, %rcx
	subq	$1, %rcx
	movq	%rcx, 40(%rsp)
	cmpq	$1, %rax
	jb	.LBB10_4
	jmp	.LBB10_3
```

# Are there any user-facing changes?

Technically, this is an API change because I made two consts `pub(crate)` instead of `pub`. However, if anyone wants access to the original consts they could just copy them into their code base.

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!---
If there are any breaking changes to public APIs, please add the `breaking change` label.
-->

andygrove commented on arrow-data/src/decimal.rs:
```diff
@@ -738,23 +738,32 @@ pub fn validate_decimal_precision(value: i128, precision: u8) -> Result<(), Arro
         )));
     }
-
-    let max = MAX_DECIMAL_FOR_EACH_PRECISION[usize::from(precision) - 1];
-    let min = MIN_DECIMAL_FOR_EACH_PRECISION[usize::from(precision) - 1];
```
I believe that this is where the `memcpy` was being introduced

viirya replied:
Hmm, this is interesting. By assigning a new variable, it did memory copying of the i128 value.

viirya commented on arrow-array/src/types.rs:
```diff
@@ -1254,6 +1260,10 @@ impl DecimalType for Decimal128Type {
         validate_decimal_precision(num, precision)
     }
+
+    fn is_valid_decimal_precision(value: Self::Native, precision: u8) -> bool {
+        is_validate_decimal_precision(value, precision)
+    }
```
Is this used to avoid creating `ArrowError` and the error string?

andygrove replied:
Yes

viirya commented on arrow-data/src/decimal.rs:
```diff
@@ -764,18 +773,28 @@ pub fn validate_decimal256_precision(value: i256, precision: u8) -> Result<(), A
 }
+
+/// Determines whether the specified `i256` value can be properly
+/// interpreted as a Decimal number with precision `precision`
```
```suggestion
/// interpreted as a Decimal256 number with precision `precision`
```

comphead reviewed (commented):
thanks @andygrove
probably a nit: I can see `usize::from(precision) - 1` a lot which require new allocation of usize.
My feeling the precision is static? may we can precompute idx and pass it by reference?

andygrove commented:
> thanks @andygrove probably a nit: I can see `usize::from(precision) - 1` a lot which require new allocation of usize. My feeling the precision is static? may we can precompute idx and pass it by reference?

What if we just insert an extra element in the start of each of the lookup arrays and then just use `precision as usize` directly as the index?

andygrove commented:
Updated benchmark results after removing the `idx` allocation:

```
null_if_overflow_precision_128
                        time:   [12.555 µs 12.591 µs 12.628 µs]
                        change: [-85.358% -85.314% -85.275%] (p = 0.00 < 0.05)
                        Performance has improved.

null_if_overflow_precision_256
                        time:   [25.614 µs 25.755 µs 25.903 µs]
                        change: [-87.983% -87.827% -87.620%] (p = 0.00 < 0.05)
                        Performance has improved.

validate_decimal_precision_128
                        time:   [74.290 ns 74.703 ns 75.109 ns]
                        change: [-7.7409% -7.2664% -6.8248%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  5 (5.00%) low mild

validate_decimal_precision_256
                        time:   [215.27 ns 216.01 ns 216.75 ns]
                        change: [-4.4231% -3.7018% -2.9993%] (p = 0.00 < 0.05)
                        Performance has improved.
```

comphead reviewed (approved):
lgtm thanks @andygrove

andygrove commented on arrow-data/src/decimal.rs:
```diff
@@ -644,8 +650,11 @@ pub(crate) const MIN_DECIMAL_BYTES_FOR_LARGER_EACH_PRECISION: [i256; 76] = [
+/// be stored in [arrow_schema::DataType::Decimal128] value of precision `p`.
+/// The first element is unused and is inserted so that we can look up using
+/// precision as the index without the need to subtract 1 first.
+pub(crate) const MAX_DECIMAL_FOR_EACH_PRECISION: [i128; 39] = [
```
Note the change from `pub` to `pub(crate)`, technically an API change.

andygrove commented:
I reverted the API change

andygrove commented:
@viirya @comphead There have been changes since this PR was approved. Please take another look when you can.

comphead commented:
I'm okay with the PR, although the array that changed moved to crate visibility and doesn't seem to affect others, however I'm not sure if anyone else other than arrow-rs actually used it

andygrove commented:
> I'm okay with the PR, although the array that changed moved to crate visibility and doesn't seem to affect others, however I'm not sure if anyone else other than arrow-rs actually used it

I reverted the change to the original array and added a copy with the extra element so as not to break the current API. The new array is `pub(crate)`

Dandandan reviewed (approved):
Nice
